xCluster: add Nodes parameter (enables cluster creation with several nodes)

woodp opened the pull request:
**Pull Request (PR) description**
Adds a "Nodes" string array parameter, specifying node names that will be added to the cluster along with the local computer at creation time (i.e passed to the New-Cluster command).

**This Pull Request (PR) fixes the following issues:**
#148, #157
The motivation behind this PR is that I was unable to join a secondary node to the cluster, tried both from the secondary node and from the primary node. The only way I had to achieve a cluster with more than one node was to specify all the nodes in the creation (using New-Cluster command and passing all the nodes into the -Node parameter).

**Task list:**
Will complete the task list if this PR makes sense to the owners.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xfailovercluster/177)
<!-- Reviewable:end -->

johlju commented:
Great work on this one! This looks good. It doesn't break the previous behavior. I'm good with this change. Please continue with the task list. :smile: 
I made a few comments, please see if they need to be addressed as well. 

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YpRy0Y8PPaYYHf_oT:-L5YpRy0Y8PPaYYHf_oU:b-djh2s8) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L46)):*
> ```PowerShell
>         [Parameter()]
>         [System.String[]]
>         $Nodes,
> ```

This parameter is not needed here since it is non-mandatory and not used in the Get-TargetResource function.
Please see https://github.com/PowerShell/DscResources/blob/master/BestPractices.md#get-targetresource-should-not-contain-unused-non-mandatory-parameters

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 202 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YmNM-xt9YaSYbb87M:-L5YmNM0IpOntoEz9Kl-:bf4rmv2) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L202)):*
> ```PowerShell
>             if($Nodes -ne $null)
>             {
>                 $AllNodes += $Nodes
> ```

If `$Nodes` already contains `$env:COMPUTERNAME` this add a duplicate? :thinking: 
Throughout.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 249 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YnVsO4dgx-yS-a4cH:-L5YnVsO4dgx-yS-a4cI:bf5vd9k) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L249)):*
> $list

Would you mind renaming this variable to `$existingNodes` or something similar if there is a better descriptive name for this instead of `list'? This is to align with the comment for the same thing in the other function. 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 258 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YofEHbC8G-qPLRb7G:-L5YofEHbC8G-qPLRb7H:bgolu4t) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L258)):*
> ```PowerShell
>                         Write-Verbose -Message ($script:localizedData.RemoveOfflineNodeFromCluster -f $targetNodeName, $Name)
> 
>                         Remove-ClusterNode -Name $targetNodeName -Cluster $Name -Force
> ```

In Test-TargetResource break was used to exit the foreach-loop. Is that appropriate here too?

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 379 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5Yo5GwuO8LibDh09r-:-L5Yo5GwuO8LibDh09r0:b-k2yb0k) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L379)):*
> $list

Would you mind renaming this variable to `$existingNodes` or something similar if there is a better descriptive name for this instead of `list'.?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-:-L5Yozk9LfSnIxpRtJd9:b-5a8cbh)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YpRy0Y8PPaYYHf_oT:-L5jMUtpTzMqHg82hyGA:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L46)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This parameter is not needed here since it is non-mandatory and not used in the Get-TargetResource function.
Please see https://github.com/PowerShell/DscResources/blob/master/BestPractices.md#get-targetresource-should-not-contain-unused-non-mandatory-parameters
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 202 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YmNM-xt9YaSYbb87M:-L5jMCe5HK9okF-6Hdg_:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L202)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

If `$Nodes` already contains `$env:COMPUTERNAME` this add a duplicate? :thinking: 
Throughout.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 249 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YnVsO4dgx-yS-a4cH:-L5jMDgBxMaYn1Z5__Vu:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L249)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> $list

Would you mind renaming this variable to `$existingNodes` or something similar if there is a better descriptive name for this instead of `list'? This is to align with the comment for the same thing in the other function. 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 258 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5YofEHbC8G-qPLRb7G:-L5jN2tN6SXUQsnEyk7B:b1c472u) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L258)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

In Test-TargetResource break was used to exit the foreach-loop. Is that appropriate here too?
</blockquote></details>

I don't think so, because we need to go through all of the nodes in the Set

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 379 at r1](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L5Yo5GwuO8LibDh09r-:-L5jMElDLw1mG03sBH5C:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L379)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> $list

Would you mind renaming this variable to `$existingNodes` or something similar if there is a better descriptive name for this instead of `list'.?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
I addressed your point, updated the mof file, fixed some styling errors and added a unit test.
Nevertheless I get several failing tests, including the one I added, and I'm unable to figure out why they are failing, I need some guidance on this.
Also appveyor complaints about some invalid characters used, I'm using "@()" to initialize an array. Should I be using a different way to create an empty array ?

woodp commented:
*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 258 at r1](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L5YofEHbC8G-qPLRb7G:-L5sPM2H7hjF9TLqAf8H:bm3t9v4) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L258)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

I don't think so, because we need to go through all of the nodes in the Set
</blockquote></details>

You are right, now is done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 258 at r1](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L5YofEHbC8G-qPLRb7G:-L5sP_vf9HbhtPL1j9w5:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/06f89162fc2dd6cbbc1f7f363c9a1135927b95ae/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L258)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

You are right, now is done
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Sorry for being away so long for this PR. It has been a lot of work at the day job and maintaining in other resources modules. But I will focus on getting this through not. :smile: 

Found a few more review comments for this one.

---

Reviewed 2 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7ytxLdos2I9KvxsKXx:-L7ytxLdos2I9KvxsKXy:b-gbijxh) (no related file):*
There are failing tests for the Set-TargetResource. Can you look them over and see why they are failing with this code change?

https://ci.appveyor.com/project/PowerShell/xfailovercluster/build/1.6.342.0#L474

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 222 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yrQxGwJxKnzjYaMIR:-L7yrQxGwJxKnzjYaMIS:b-8g6mmp) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L222)):*
> Get-All-Nodes($Nodes)

Please change this to named arguments `Get-AllNodes -Nodes $Nodes`. Throughout.

*Also see other comment for renaming this function.*

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 222 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yrtPPgdJODw4lZEs4:-L7yrtPPgdJODw4lZEs5:b-3itcfg) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L222)):*
> $AllNodes

Please rename this to use camelCase syntax (`$allNodes` lower case 'a'). Throughout.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 233 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yqoLZ8UE8-_Lxis7J:-L7yqoLZ8UE8-_Lxis7K:bn9lcuk) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L233)):*
> if(

Please add a space between the `if`-statement and the parenthesis 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 365 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7ysjKHWD5yCzWTT9kK:-L7ysjKHWD5yCzWTT9kL:b-nl4wxy) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L365)):*
> foreach(

Please add a space between the `foreach`-statement and the parenthesis 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 376 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yt1wekc6WlIGBhcyH:-L7yt1wekc6WlIGBhcyI:b51iqe4) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L376)):*
> ```PowerShell
>                         {
>                             Write-Verbose -Message ($script:localizedData.ClusterNodeIsDown -f $targetNodeName, $Name)
>                             $returnValue = $false
> ```

Should this be `$found = $false`? For the evaluation on row 383 to be true? 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 383 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yso7UKwfJLan_W2_k:-L7yso7UKwfJLan_W2_l:bn9lcuk) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L383)):*
> if(

Please add a space between the `if`-statement and the parenthesis 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 530 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yqyB8kDbGwGvyXT8h:-L7yqyB9zGBIRBSMWFVL:b-gg4ewk) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L530)):*
> Get-All-Nodes

Could we use the Verb-Noun syntax for this? So for this function name it would be `Get-AllNodes` instead?

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 543 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yrD0NTABpuXHxXDrM:-L7yrD0NTABpuXHxXDrN:bn9lcuk) ([raw file](https://github.com/powershell/xfailovercluster/blob/82aa0b8dc589db417c4068503a89a99c3de20760/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L543)):*
> if(

Please add a space between the `if`-statement and the parenthesis 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-:-L7ytTSDda09hG9gMbSc:b-aq9mkw)*
<!-- Sent from Reviewable.io -->

johlju commented:
I did a typo that could have been seen as strange, what I meant to say in previous review comment was that I will focus on getting this through now. Let me know if you need any help 😄

woodp commented:
*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 376 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yt1wekc6WlIGBhcyH:-L7zW6oo1sFNUgX26vrO:b5l189c) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L376)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Should this be `$found = $false`? For the evaluation on row 383 to be true? 
</blockquote></details>

not sure about this one

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 222 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yrQxGwJxKnzjYaMIR:-L7zVzkG32rKlZdzgpUj:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L222)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> Get-All-Nodes($Nodes)

Please change this to named arguments `Get-AllNodes -Nodes $Nodes`. Throughout.

*Also see other comment for renaming this function.*
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 222 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yrtPPgdJODw4lZEs4:-L7zW-vG6-7VRUoaoDvE:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L222)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> $AllNodes

Please rename this to use camelCase syntax (`$allNodes` lower case 'a'). Throughout.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 233 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yqoLZ8UE8-_Lxis7J:-L7zW1mHDx1G5mGS2TQ5:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L233)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> if(

Please add a space between the `if`-statement and the parenthesis 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 365 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7ysjKHWD5yCzWTT9kK:-L7zW5AtAGoAR8hYxomq:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L365)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> foreach(

Please add a space between the `foreach`-statement and the parenthesis 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 383 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yso7UKwfJLan_W2_k:-L7zWv0X6X6YQ_MlWKBs:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L383)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> if(

Please add a space between the `if`-statement and the parenthesis 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 530 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yqyB8kDbGwGvyXT8h:-L7zWwjGBMI5cy_l18Ib:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L530)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> Get-All-Nodes

Could we use the Verb-Noun syntax for this? So for this function name it would be `Get-AllNodes` instead?
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 543 at r3](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7yrD0NTABpuXHxXDrM:-L7zVw198yzV2W3MG2iM:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L543)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> if(

Please add a space between the `if`-statement and the parenthesis 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 376 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yt1wekc6WlIGBhcyH:-L7zjC4U00yLzH7o79ll:b-akth8t) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L376)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

not sure about this one
</blockquote></details>

LGTM. This is good as-is. Was me looking at it wrongly. Thanks for correcting me.

---

*[Tests/Unit/MSFT_xCluster.Tests.ps1, line 322 at r2](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7zgjwbCrha99PB51qW:-L7zgjwcPBSYj8qBnmbF:b-ozuc0d) ([raw file](https://github.com/powershell/xfailovercluster/blob/4938f91ba540849b3d24ef57b460813f52a599d7/Tests/Unit/MSFT_xCluster.Tests.ps1#L322)):*
> Assert-MockCalled -CommandName New-Cluster -Exactly -Times 1 -Scope It -ParameterFilter {
>                                     $Nodes -eq 'foo' -and
>                                     $Nodes -eq 'bar'
>                                 }

I liked this `ParameterFilter` to assert that the call to `New-Cluster` cmdlet was the correct one. But maybe this can work for the parameter filter?
```
Assert-MockCalled -CommandName New-Cluster -Exactly -Times 1 -Scope It -ParameterFilter {
    $Node.Contains($env:COMPUTERNAME) -eq $true `
    -and $Node.Contains('foo') -eq $true `
    -and $Node.Contains('bar) -eq $true
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 376 at r3](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L7yt1wekc6WlIGBhcyH:-L7zvfLEKyIVOVlvzd-t:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/2e260bbbbc30fa3284c0487e15a6c22a490c440d/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L376)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

LGTM. This is good as-is. Was me looking at it wrongly. Thanks for correcting me.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/MSFT_xCluster.Tests.ps1, line 322 at r2](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7zgjwbCrha99PB51qW:-L82UBa12gkd3035gE7O:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/4938f91ba540849b3d24ef57b460813f52a599d7/Tests/Unit/MSFT_xCluster.Tests.ps1#L322)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> Assert-MockCalled -CommandName New-Cluster -Exactly -Times 1 -Scope It -ParameterFilter {
>                                     $Nodes -eq 'foo' -and
>                                     $Nodes -eq 'bar'
>                                 }

I liked this `ParameterFilter` to assert that the call to `New-Cluster` cmdlet was the correct one. But maybe this can work for the parameter filter?
```
Assert-MockCalled -CommandName New-Cluster -Exactly -Times 1 -Scope It -ParameterFilter {
    $Node.Contains($env:COMPUTERNAME) -eq $true `
    -and $Node.Contains('foo') -eq $true `
    -and $Node.Contains('bar) -eq $true
}
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[a discussion](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7ytxLdos2I9KvxsKXx:-L82UHOKBSyUrCq12bBD:bb7f2mf) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

There are failing tests for the Set-TargetResource. Can you look them over and see why they are failing with this code change?

https://ci.appveyor.com/project/PowerShell/xfailovercluster/build/1.6.342.0#L474
</blockquote></details>

done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[Tests/Unit/MSFT_xCluster.Tests.ps1, line 322 at r2](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L7zgjwbCrha99PB51qW:-L82UR606CGFdJh42245:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/4938f91ba540849b3d24ef57b460813f52a599d7/Tests/Unit/MSFT_xCluster.Tests.ps1#L322)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L833TlevIZG-loMBPJo:-L833TlevIZG-loMBPJp:bh6jyv9) (no related file):*
Please add the `Node`parameter to the resource in the README.md.

*See other comment of renaming `Nodes` to  `Node`.* 

---

*[a discussion](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L833bMddSS9dx2NdtMs:-L833bMddSS9dx2NdtMt:bijkwkr) (no related file):*
Please update the Unreleased section in the CHANGELOG.md with an entry explaining this change. 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 83 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L832DRW8IT_mky4meHz:-L832DRW8IT_mky4meI-:bukgs1s) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L83)):*
> ```PowerShell
>     }
> 
>     @{
> ```

We should return the existing node names in the property `Node` in this hash table.

*See other comment of renaming `Nodes` to  `Node`.* 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 98 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L832Rlpmr-xx_g3Bnwn:-L832Rlpmr-xx_g3Bnwo:b18v60o) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L98)):*
> Nodes

I suggest we rename this to `Node` singular, because this can be set to one or more.  It aligns with the other parameters. Throughout. 

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 539 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L83004mMVwj4C9FSK6C:-L83004nI1fkawnw-r3Z:b9mv2r3) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L539)):*
> ```PowerShell
> 
>     $allNodes = @()
>     $allNodes += $env:COMPUTERNAME
> ```

A question. It is not really clear in the code that $Nodes will always default to `$env:COMPUTERNAME`, so I thought of the following.

Would it work to add `$env:COMPUTERNAME` as the default value for `$Nodes` parameter, and remove the entry from here?  Wouldn't the user specify the current target node and any other nodes in this parameter, and if it is not assigned it will be set to the default value`$env:COMPUTERNAME`? If so, is this helper function necessary? 

If wer add the value as default value to the `$Nodes` parameter, we also need to update the parameter description in README.md, comment-based help and schema.mof.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 224 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L8353UlDVHCsNSJ0XEL:-L8353UlDVHCsNSJ0XEM:blmnhtw) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L224)):*
> ```PowerShell
>             Write-Verbose -Message ($script:localizedData.AddNodeToCluster -f $targetNodeName, $Name)
> 
>             $existingNodes = Get-ClusterNode -Cluster $Name
> ```

is this scenario supported? 

`Node` is set to `@('Server1','Server2')`. Existing nodes is reported as `@('Server1','Server2','Server3')`. User did not specify the last node 'Server3', because the user wanted it removed from the cluster.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 539 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L83004mMVwj4C9FSK6C:-L83l2y0K1h7gJhFZBcO:b-kp4w12) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L539)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

A question. It is not really clear in the code that $Nodes will always default to `$env:COMPUTERNAME`, so I thought of the following.

Would it work to add `$env:COMPUTERNAME` as the default value for `$Nodes` parameter, and remove the entry from here?  Wouldn't the user specify the current target node and any other nodes in this parameter, and if it is not assigned it will be set to the default value`$env:COMPUTERNAME`? If so, is this helper function necessary? 

If wer add the value as default value to the `$Nodes` parameter, we also need to update the parameter description in README.md, comment-based help and schema.mof.
</blockquote></details>

That sound good to me, although the helper function also removes the duplicate objects, so I think we would need to leave it anyway

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 224 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L8353UlDVHCsNSJ0XEL:-L83lYptPVAf_Fkq8rrO:bboil4q) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L224)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

is this scenario supported? 

`Node` is set to `@('Server1','Server2')`. Existing nodes is reported as `@('Server1','Server2','Server3')`. User did not specify the last node 'Server3', because the user wanted it removed from the cluster.

</blockquote></details>

No, I didn't consider that scenario. Will think it over.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 539 at r5](https://reviewable.io:443/reviews/powershell/xfailovercluster/177#-L83004mMVwj4C9FSK6C:-L87sPHK8pS3iPmqty7W:bh0oi32) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L539)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

That sound good to me, although the helper function also removes the duplicate objects, so I think we would need to leave it anyway
</blockquote></details>

Sounds good filtering out unique nodes names. But maybe we should rename the helper function to `Get-UniqueNodeName` (or plural `...Names`)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 539 at r5](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L83004mMVwj4C9FSK6C:-L8IhdYg3Ri8XaVf2dhT:bl1je3i) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L539)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Sounds good filtering out unique nodes names. But maybe we should rename the helper function to `Get-UniqueNodeName` (or plural `...Names`)?
</blockquote></details>

after removing that it was reduced to a one line function, so I just inlined it

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
Review status: 1 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 83 at r5](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L832DRW8IT_mky4meHz:-L8IhS5G5oS4ihTENpkO:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L83)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should return the existing node names in the property `Node` in this hash table.

*See other comment of renaming `Nodes` to  `Node`.* 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 98 at r5](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L832Rlpmr-xx_g3Bnwn:-L8IhUas--FvS9wGjOoZ:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L98)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> Nodes

I suggest we rename this to `Node` singular, because this can be set to one or more.  It aligns with the other parameters. Throughout. 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCluster/MSFT_xCluster.psm1, line 224 at r5](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L8353UlDVHCsNSJ0XEL:-L8IhY_FE80bkz1mytnZ:b-896fix) ([raw file](https://github.com/powershell/xfailovercluster/blob/f6aedf018bd3aaff10f7f484c6c89e434f3e8f4c/DSCResources/MSFT_xCluster/MSFT_xCluster.psm1#L224)):*
<details><summary><i>Previously, woodp (Pedro Wood) wrote…</i></summary><blockquote>

No, I didn't consider that scenario. Will think it over.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[a discussion](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L833TlevIZG-loMBPJo:-L8XKYfo2-LEJBuylFfs:bb7f2mf) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please add the `Node`parameter to the resource in the README.md.

*See other comment of renaming `Nodes` to  `Node`.* 
</blockquote></details>

done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

woodp commented:
*[a discussion](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177#-L833bMddSS9dx2NdtMs:-L8XK_gT0ahkF6RBKoHz:bb7f2mf) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please update the Unreleased section in the CHANGELOG.md with an entry explaining this change. 
</blockquote></details>

done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/powershell/xfailovercluster/177)*
<!-- Sent from Reviewable.io -->

johlju commented:
Labeling this PR as abandoned since it has gone 14 days or more since the last update. An abandoned PR can be continued by another contributor. The abandoned label will be removed if work on the PR is taken up again.

woodp commented:
*[a discussion](https://reviewable.io/reviews/powershell/xfailovercluster/177#-L8Y-DOs6Tv1eyxxtA6v:-LHERFcK8hGrLH2sOCGu:bo0qr4b) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

The tests are failing, can you look and see if you can resolve them?
</blockquote></details>

I'm not being able to figure out why these 2 tests fail. For the first one I always get an exception, if I add the filter for the GetClusterNode mock, I get "StubNotImplemented" otherwise I get an exception without any info at all. Is very difficult with only this info to know why it fails. Also don't seem to get how the mocks/stubs work.

---


<!-- Sent from Reviewable.io -->

johlju commented:
The Stubs are used to have something to attach the mock to. A unit test should not need the actual module with the cmdlets installed, instead those are replaced by stubs (empty functions except for the throw you are seeing) so that when using `Mock` the mock can find the function and get the parameters from the stub. 
If the stub is not mocked, or mocked with the wrong parameter filter, the stub throws a not implemented error so that we can know the mock was not called but the actual function (the stub or the cmdlet in the Failover Cluster module). 

So in your case, when you get StubNotImplemented the mock is not called. 

Did this help you? Let me know if you need more information.

johlju commented:
Closing this PR due to inactivity. If there are someone who wants to work on this issue or PR please open a new PR.
